%\VignetteIndexEntry{Bpest data}
%\VignetteDepends{Bpest}
%\VignettePackage{Bpest}

\documentclass[12pt]{article}

\usepackage{Sweave}

\textwidth=6.2in
\textheight=8.5in
\oddsidemargin=0.2in
\evensidemargin=0.2in
\headheight=0in
\headsep=0in

\begin{document}
\SweaveOpts{concordance=TRUE}

\title{Bpest}
\author{Conrad Burden, Sylvain For\^et, Peijie Lin}
\date{\today}
\maketitle

Many of the known mechanisms driving gene regulation fall into the category of epigenomic modifications.
DNA methylation is a common epigenomic modification, in which a cytosine (C) in the genomic DNA sequence can be altered by the addition of a methyl group.
Methylation patterns can be are detected by treating DNA with bisulphite, which converts unmethylated cytosines to uracils while leaving methylated cytosines intact.
This can be carried out at specific loci in a genome by amplifying fragments of interest by PCR.
These PCR products (called amplicons) can be mapped to a reference and methylation patterns inferred. 

However, the bisulphite conversion is not 100\% efficient, and this introduces errors in the observed distribution of methylation patterns.
A second source of errors is the sequencing error.
{\tt Bpest} (for \textbf{B}isulphite \textbf{p}attern \textbf{est}imator) calculates the estimated distribution over methylation patterns based on an input of methylation pattern count data, an incomplete conversion rate and site-dependent read error rates.

The main component of the package is the function {\tt estimatePatterns()}, which generates a table of estimates $\hat\theta_i$ of the distribution over methylation patterns, a plot that compares the observed read distribution with the estimated distribution, and a list of patterns identified as spurious.
Input to the function is a data frame {\tt patternCounts} listing the methylation patterns in the first column followed by a number of columns of count data (one column per sample).
Estimation will be performed on the column specified by the variable {\tt column} using the non-conversion and error rates specified by the parameters {\tt epsilon} and {\tt eta} respectively.
The boolean variable {\tt fast} enables either a fast implementation (default) which ignores those patterns for which the observed read count is zero or a slow impementation.
The parameter {\tt steps} is passed to the function {\tt constrOptim()} to control the accuracy of the determination of the maximum log-likelihood, and {\tt yLimit1} and  {\tt yLimit2} control the range of the y-axis on the plots produced.

In the following example, the input is the table of counts {\tt patternsExample}.
We haven't specified the {\tt column} variable so the function {\tt estimatePatterns()} by default applies to the first column of counts.
The parameter {\tt epsilon} is 0.01, while the parameter {\tt eta} is not specified and by default is 0. 
Two plots are produced: the lower plot is the expanded version of the upper plot.

<<>>=
library(Bpest)
data(patternsExample)
patternsExample
estimatePatterns(patternsExample, epsilon=0.01)
@

In the following example we still use the same input table but instead now specify {\tt column = 2}. 
We set variables {\tt epsilon=0.005}, {\tt eta=c(0.008, 0.01, 0.01, 0.01, 0.008)} (denoting different error rate at each site) and {\tt yLimit2=0.04}.
We also turn off the plotting.

<<>>=
estimatePatterns(patternsExample,
                 epsilon=0.01, 
                 eta=c(0.008, 0.01, 0.01, 0.01, 0.008), 
                 column=2)
@

Note that in this example four patterns have been identified as spurious; they are patterns {\tt 01010}, {\tt 10001}, {\tt 10010} and {\tt 11000}.

\end{document}
